src/detect-engine-alert.c

```diff
@@ -337,72 +336,141 @@ static inline void FlowApplySignatureActions(
-    qsort(det_ctx->alert_queue, det_ctx->alert_queue_size, sizeof(PacketAlert),
-            AlertQueueSortHelper);
+        /* set actions on the flow */
+        FlowApplySignatureActions(p, pa, s, pa->flags);
 
-    uint16_t i = 0;
-    uint16_t max_pos = det_ctx->alert_queue_size;
+        /* set actions on packet */
+        PacketApplySignatureActions(det_ctx, p, s, pa->flags);
+    }
 
-    while (i < max_pos) {
-        const Signature *s = de_ctx->sig_array[det_ctx->alert_queue[i].num];
-        int res = PacketAlertHandle(de_ctx, det_ctx, s, p, &det_ctx->alert_queue[i]);
-
-        if (res > 0) {
-            /* Now, if we have an alert, we have to check if we want
-             * to tag this session or src/dst host */
-            if (s->sm_arrays[DETECT_SM_LIST_TMATCH] != NULL) {
-                KEYWORD_PROFILING_SET_LIST(det_ctx, DETECT_SM_LIST_TMATCH);
-                SigMatchData *smd = s->sm_arrays[DETECT_SM_LIST_TMATCH];
-                while (1) {
-                    /* tags are set only for alerts */
-                    KEYWORD_PROFILING_START;
-                    sigmatch_table[smd->type].Match(det_ctx, p, (Signature *)s, smd->ctx);
-                    KEYWORD_PROFILING_END(det_ctx, smd->type, 1);
-                    if (smd->is_last)
-                        break;
-                    smd++;
-                }
-            }
+    /* Thresholding removes this alert */
+    if (res == THRESHOLD_SUPPRESSED || res == THRESHOLD_SILENT_MATCH ||
+            (s->flags & SIG_FLAG_NOALERT)) {
+        /* we will not copy this to the AlertQueue */
+        return THRESHOLD_SUPPRESSED;
+    } else {
+        return THRESHOLD_NOT_SUPPRESSED;
+    }
+}
 
-            /* set actions on the flow */
-            FlowApplySignatureActions(
-                    p, &det_ctx->alert_queue[i], s, det_ctx->alert_queue[i].flags);
+/**
+ * \brief Iterate over the packet alerts queue, removing those that didn't match
+ *        the threshold, and those that match after a signature with the action
+ *        "pass". The queue is sorted by action priority/order
+ *
+ * \param de_ctx detection engine context
+ * \param det_ctx detection engine thread context
+ * \param p pointer to the packet
+ */
+void PacketAlertQueueFinalize(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx, Packet *p)
+{
+    SCEnter();
 
-            /* set actions on packet */
-            PacketApplySignatureActions(p, s, det_ctx->alert_queue[i].flags);
-        }
+    uint16_t max_pos = det_ctx->alert_queue_size;
 
-        /* Thresholding removes this alert */
-        if (res == 0 || res == 2 || (s->flags & SIG_FLAG_NOALERT)) {
-            /* we will not copy this to the AlertQueue */
-            p->alerts.suppressed++;
-        } else if (p->alerts.cnt < packet_alert_max) {
-            p->alerts.alerts[p->alerts.cnt] = det_ctx->alert_queue[i];
+    /* If there are alerts for this packet, process them */
+    if (max_pos > 0) {
```

Comment:
I have noticed that we call PacketAlertQueueFinalize even when there are no alerts in the packet, so I'm thinking maybe we can optimize that, too.

Did that here, also because the CI Fuzz found a bug caused when max_pos is zero (no alerts for packet).

Reply:
Worth looking into, but lets do that after this work has been finalized to not intro more moving parts.

Reply:
roger!